### DISCUSSION TITLE
I2C read data without global variables

### DISCUSSION THREAD

**Discussion by ben-jcp** [MrYsLab/telemetrix#40] [Q&A] [answered]
Hi there,

Admittedly somewhat of a novice to Python here, so please forgive me for missing anything obvious! I'm trying to use Telemetrix to read data from a sensor (MS5837-30BA) via an Arduino UNO using I2C for a project. I'd like the `i2c_read()` command to return the content of the sensor's response as the output of a function, and not just print it as in examples I've seen. I was using the `i2c_get_read_data` command in PyMata which returned the data easily, but too slowly. The only way I can work out to extract the data is by using a global variable in the callback, which is obviously undesirable:
```
def callbackP(data):
    global Pdata # define a global to store the data returned in I2C read
    Pdata = data

def read_sequence(sensor_address = sensor_address):
    board.i2c_write(sensor_address, [0x4A]) # 0x4A is command to sensor to write D1 to ADC
    sleep_ms(20) # datasheet says write to ADC takes up to 18.08ms
    board.i2c_read(sensor_address,0,3,callbackP) # reads ADC value (3 bytes)
    sleep_ms(200) # sleep while callback function runs – otherwise None is returned
    D1 = 65536*Pdata[5] + 256*Pdata[6] + Pdata[7] # storing combined value of data bytes

def sleep_ms(duration, get_now=time.perf_counter): # sleep function for improved accuracy to time.sleep()
    now = get_now()
    end = now + duration/1000
    while now < end:
        now = get_now()
```

Is there any way I can get the data out more easily?  And is there a way to speed up the I2C reading of the data (in either package; here would be avoiding the 200ms delay) – I've verified with an oscilloscope that the sensor responds very quickly, so I imagine the holdup is in the Arduino-Python interface?

Thanks in advance for any help :)

**Comment by MrYsLab** [comment, marked as answer]
Hi, thanks for the question. I have no experience with this sensor. However, have you tried using an example from an Arduino sketch library such as the [BlueRobotics MS5837 Library](https://github.com/bluerobotics/BlueRobotics_MS5837_Library/tree/master/examples/MS5837_Example)? 
This example contains a 1-second delay between reads, so I suggest trying to shorten this delay and see the fastest response you can get. This will eliminate Telemetrix from the equation and provide a good baseline for the speed of a read.

When confronted with having to deal with globals, I typically restructure my code as a class to avoid the whole global mess. You may still choose to use a procedural model with the global, but what I would do is remove the 200 ms. delay. Create D1 as a global value, initialize its value to zero,  and set its value in the callback function. You would then poll the value of D1 to see if it is non-zero, and if it is, retrieve its value and reset D1 to zero for the next read.

If you need to run in a non-blocking mode, you could create a separate thread to do the polling of D1. Or, if you are adventurous, you can use asyncio instead of threading. 

Please let me know if you have any other questions. Thanks again for the question.

**Reply by ben-jcp** [reply]
Hi, thank you for your reply (and for writing this great library)!

I've tried using the example sketch, it's happy to return readings with a 30ms delay in the code, which is more than fast enough. It seems to be taking only ~39ms for the reads to be executed (for a total of ~70ms). This is roughly twice the 18ms max wait for the sensor to carry out the ADC conversion, which makes sense as there are two data values to be converted and sent in each read. So it seems like the slowing is on the Python side of things – perhaps there's no way to avoid this.

As for the global variable issue, using classes is something that I will look into, thank you. Just so I'm sure I understand it correctly, am I right in thinking that `i2c_read()` will do the comms via I2C, then execute the specified callback function with the data list as its argument?

Thanks again for your help!

**Reply by MrYsLab** [reply]
Thanks, I am glad you are enjoying the library. 
Yes, you are correct. When you issue an i2c_read command on the Python side, the Arduino code will do the [i2c read](https://github.com/MrYsLab/Telemetrix4Arduino/blob/c6a805917116b687b90182b83d627670f8ac27d0/examples/Telemetrix4Arduino/Telemetrix4Arduino.ino#L940) and return the data as a callback. 
Just make sure that you have called set_pin_mode_i2c before doing the reads. You only need to set the pin mode once.
If you have any other questions, I would be pleased to try to provide answers.
